polyply/src/top_parser.py

```diff
@@ -308,10 +308,33 @@ def _atomtypes(self, line, lineno=0):
         """
         Parse and store atomtypes section
         """
-        atom_name = line.split()[0]
-        nb1, nb2 = line.split()[-2:]
-        self.topology.atom_types[atom_name] = {"nb1": float(nb1),
-                                               "nb2": float(nb2)}
+        atom_type_line = {"atom_num": None,
+                          "bond_type": None,
+                          "mass": None,
+                          "charge": None,
+                          "ptype": None,
+                          "nb1": None,
+                          "nb2": None}
+
+        tokens = line.split()
+        atom_name = tokens.pop(0)
+        for var in ["nb2", "nb1", "ptype", "charge", "mass"]:
+             if var != "ptype":
+                atom_type_line[var] = float(tokens.pop(-1))
+             else:
+                atom_type_line[var] = tokens.pop(-1)
+
+        if len(tokens) >= 1:
+            atom_type_line["atom_num"] = float(tokens.pop(-1))
+
+        if len(tokens) == 1:
+            atom_type_line["bond_type"] = tokens.pop(0)
```

Comment:
```suggestion
        from itertools import zip_longest
        atom_type_line = dict(zip_longest(["atom_name", "nb2", "nb1", "ptype", "charge", "mass"], tokens)
        if None in atom_type_line:
            msg = ("Can't parse line {}. Found more parameters than expected.")
            raise OSError(msg.format(line))
```
Something along those lines anyway. Maybe zip with a defaultdict is better than zip_longest.

Reply:
I think I captured the essence of what you suggested.